source/date_time.c

```diff
@@ -0,0 +1,826 @@
+                case 't':
+                    c = str[++index];
+                    switch (c) {
+                        case 'C':
+                        case 'c':
+                            return true;
+                        default:
+                            return false;
+                    }
+
+                case 'C':
+                case 'c':
+                    c = str[++index];
+                    switch (c) {
+                        case 'T':
+                        case 't':
+                            return true;
+                        default:
+                            return false;
+                    }
+                default:
+                    return false;
+            }
+        case 'G':
+        case 'g':
+            c = str[++index];
+            switch (c) {
+                case 'M':
+                case 'm':
+                    c = str[++index];
+                    switch (c) {
+                        case 'T':
+                        case 't':
+                            return true;
+                        default:
+                            return false;
+                    }
+                default:
+                    return false;
+            }
+        case '+':
+        case '-':
+            c = str[++index];
+            switch (c) {
+                case '0':
+                    c = str[++index];
+                    switch (c) {
+                        case '0':
+                            c = str[++index];
+                            switch (c) {
+                                case '0':
+                                    return true;
+                                default:
+                                    return false;
+                            }
+                        default:
+                            return false;
+                    }
+                default:
+                    return false;
+            }
+        case 'Z':
+            return true;
+        default:
+            return false;
+    }
+}
+
+void aws_date_time_init_now(struct aws_date_time *dt) {
+    uint64_t current_time = 0;
+    aws_sys_clock_get_ticks(&current_time);
+    dt->timestamp = (time_t)aws_timestamp_convert(current_time, AWS_TIMESTAMP_NANOS, AWS_TIMESTAMP_SECS, NULL);
+}
+
+void aws_date_time_init_epoch_millis(struct aws_date_time *dt, uint64_t ms_since_epoch) {
+    dt->timestamp = (time_t)(ms_since_epoch / AWS_TIMESTAMP_MILLIS);
+}
+
+void aws_date_time_init_epoch_secs(struct aws_date_time *dt, double sec_ms) {
+    dt->timestamp = (time_t)sec_ms;
+}
+
+static int s_parse_iso_8601(const struct aws_byte_buf *date_str, struct tm *parsed_time) {
+    size_t index = 0;
+    size_t state_start_index = 0;
+    const int final_state = 7;
+    int state = 0;
+    bool error = false;
+
+    while (state <= final_state && !error && index < date_str->len) { /* lgtm [cpp/constant-comparison] */
+        char c = date_str->buffer[index];
+        switch (state) {
```

Comment:
Please use a lexer generator. Validating these state machines by hand is not trivial.

Reply:
see above.... I'm not pulling in a dependency for this.

Reply:
Then at least write it in a more readable way. With a few token-extraction helpers you can write simple straight-line code to do the parse, something like this:

```
char tmpbuf[16];

if (!extract_token(tmpbuf, sizeof(tmpbuf), &input_ptr) {
  return -1;
}
day_of_week = parse_day_of_week(tmpbuf);

if (*input_ptr != ',') return -1;
input_ptr++;

while (*input_ptr == ' ') input_ptr++;

if (!extract_int(&day_of_month, &input_ptr)) {
  return -1;
}
// ...
```

Reply:
will do.

Reply:
I wrote it, and I hate it. It's not clear, and it doesn't have the advantage that an FSM has, which is it's strict by default. I'm going to keep these. It's a well established pattern and is what a regex or something would compile down to anyways.

Reply:
https://paste.amazon.com/show/henso/1543545571